### ISSUE TITLE
Add automated tests for fake patterns

### ISSUE THREAD

**Issue by ST-DDT** [faker-js/faker#3325]
### Clear and concise description of the problem

If we add or change fake patterns (or have to review them), we have to manually check for all possible combinations of patterns.

### Suggested solution

Add an automated test that checks fake patterns for errors including errors in nested fake patterns.

### Alternative

Keep testing by hand.

### Additional context

We have to find a way to detect all fake patterns even if they aren't suffixed with pattern.

Found in: https://github.com/faker-js/faker/pull/3324/commits/bfbd0b608fdcf05966cc1f286d315b9d5f7ac26e

- https://github.com/faker-js/faker/pull/3324/commits/bfbd0b608fdcf05966cc1f286d315b9d5f7ac26e

**Comment by matthewmayer** [comment]
Sounds useful; though what errors specifically would you test for? 

The typo i made could have been caught by checking for a balanced number of } and { characters, but not sure how common that would be.

**Comment by atzzCokeK** [comment]
@ST-DDT @matthewmayer Hi! I'd like to work on this issue if this issue is still alive. I did some investigation and benchmarking.

Pattern detection: Recursively walk all locale definition values and collect every string containing {{. This finds 472 unique patterns across all locales, regardless of file naming.

## Two types of patterns exist:

- 431 patterns use {{module.method}} syntax, resolved via helpers.fake()
- 41 patterns use {{keyword}} syntax, resolved via helpers.mustache() with a manual data map

The mustache()-based patterns are found in two places:
- hacker.phrase — uses {{verb}}, {{noun}}, etc. This could potentially be refactored to fake()-style ({{hacker.verb}}) as a separate effort.
- person.fullName — uses {{person.firstName}}, {{person.lastName}}, etc. This might not be replaced with fake() easily because it injects caller-provided values (e.g. a specific firstName or sex passed as options).

## Proposed test (two layers):

1. {/} balance check on all patterns (~1.4ms) — catches brace typos like #3324
faker.helpers.fake() execution on the 431 fake-style patterns, per locale (~21ms) — also catches invalid method refs and bad args
2. The mustache-style patterns would only get layer 1 for now. Both layers together run in under 25ms — negligible CI impact.

Does this approach sound good?

**Comment by atzzCokeK** [comment]
In hacker.phrase, the patterns could be refactored from `{{verb}}` to `{{hacker.verb}}` , which would unify the resolution method. Would a separate PR for that be welcome? If an issue would be preferred, I'm happy to create one instead.

**Comment by matthewmayer** [comment]
> In hacker.phrase, the patterns could be refactored from `{{verb}}` to `{{hacker.verb}}` , which would unify the resolution method. Would a separate PR for that be welcome? If an issue would be preferred, I'm happy to create one instead.

sure a seperate PR for that would be good

**Comment by atzzCokeK** [comment]
@matthewmayer Thank you for your reply:> I will work on that!

**Comment by matthewmayer** [comment]
> [@ST-DDT](https://github.com/ST-DDT) [@matthewmayer](https://github.com/matthewmayer) Hi! I'd like to work on this issue if this issue is still alive. I did some investigation and benchmarking.
> 
> Pattern detection: Recursively walk all locale definition values and collect every string containing {{. This finds 472 unique patterns across all locales, regardless of file naming.
> 
> ## Two types of patterns exist:
> * 431 patterns use {{module.method}} syntax, resolved via helpers.fake()
> * 41 patterns use {{keyword}} syntax, resolved via helpers.mustache() with a manual data map
> 
> The mustache()-based patterns are found in two places:
> 
> * hacker.phrase — uses {{verb}}, {{noun}}, etc. This could potentially be refactored to fake()-style ({{hacker.verb}}) as a separate effort.
> * person.fullName — uses {{person.firstName}}, {{person.lastName}}, etc. This might not be replaced with fake() easily because it injects caller-provided values (e.g. a specific firstName or sex passed as options).
> 
> ## Proposed test (two layers):
> 1. {/} balance check on all patterns (~1.4ms) — catches brace typos like [feat(internet): update to simplified modern user-agent list #3324](https://github.com/faker-js/faker/pull/3324)
>    faker.helpers.fake() execution on the 431 fake-style patterns, per locale (~21ms) — also catches invalid method refs and bad args
> 2. The mustache-style patterns would only get layer 1 for now. Both layers together run in under 25ms — negligible CI impact.
> 
> Does this approach sound good?

In general this sounds fine. To clarify, these tests dont currently catch any errors, but they would catch previously fixed errors in the locale files?

**Comment by atzzCokeK** [comment]
@matthewmayer Right, these tests don't currently catch any errors — all patterns pass. The value is in preventing regressions when locale data is added or modified.

If the hacker.phrase refactor PR is merged, a single faker.helpers.fake() execution test would cover all patterns — including brace balance, invalid method refs, and bad args. (Now I think that not necessary Layer1 anymore.) Something like:

```ts
function collectFakePatterns(data: unknown): string[] {
  if (Array.isArray(data)) {
    return data.flatMap((e) => collectFakePatterns(e));
  } else if (typeof data === 'object' && data != null) {
    return Object.values(data).flatMap((v) => collectFakePatterns(v));
  } else if (typeof data === 'string' && data.includes('{{')) {
    return [data];
  }
  return [];
}

  it('should resolve all fake patterns without errors', () => {
    for (const [locale, data] of Object.entries(allLocales)) {
      const faker = new Faker({ locale: [data, en, base] });
      for (const pattern of collectFakePatterns(data)) {
        expect(
          () => faker.helpers.fake(pattern),
          `[${locale}] ${pattern}`
        ).not.toThrow();
      }
    }
  });
```

I'll send the refactor PR first, then follow up with the test PR.

**Comment by ST-DDT** [comment]
This will be more thoroughly solved by #3176
